### PR TITLE
Make knowhere run on old cpus which only support sse4.2

### PR DESCRIPTION
**Pull request by chasingegg**
#140 
/kind bug

### CODE REVIEW

**Comment by alexanderguzhva** [comment]
1. It is not IndexScaNN being unsupported, it is PQ fast scan procedure being unsupported. Technically, there are multiple indices in Faiss that rely on PQ fast scan, so I would change `support_scann` variable name to `support_pq_fast_scan` to make it clear and acceptable for future possible indices without refactoring. 
2. Intel(R) Xeon(R) CPU E5-2420 supports AVX, but not AVX2, which is written in the flags in https://github.com/milvus-io/milvus/issues/27552 and also mentioned on the [manufacturer web page](https://www.intel.com/content/www/us/en/products/sku/64617/intel-xeon-processor-e52420-15m-cache-1-90-ghz-7-20-gts-intel-qpi/specifications.html) . So, I would transition the change into `Make knowhere run on old cpus which only support AVX`

**Comment by chasingegg** [comment]
> 1. It is not IndexScaNN being unsupported, it is PQ fast scan procedure being unsupported. Technically, there are multiple indices in Faiss that rely on PQ fast scan, so I would change `support_scann` variable name to `support_pq_fast_scan` to make it clear and acceptable for future possible indices without refactoring.
> 2. Intel(R) Xeon(R) CPU E5-2420 supports AVX, but not AVX2, which is written in the flags in [[Bug]: Milvus2.3.1 standalone exited with code 132 milvus-io/milvus#27552](https://github.com/milvus-io/milvus/issues/27552) and also mentioned on the [manufacturer web page](https://www.intel.com/content/www/us/en/products/sku/64617/intel-xeon-processor-e52420-15m-cache-1-90-ghz-7-20-gts-intel-qpi/specifications.html) . So, I would transition the change into `Make knowhere run on old cpus which only support AVX`, if possible

Actually the lowest level we are supporting is SSE4.2, I would suggest changing to `Make knowhere run on old cpus which only support SSE4.2`

File: include/knowhere/expected.h [hunk 1]
```diff
@@ -80,6 +81,8 @@ Status2String(knowhere::Status status) {
             return "raft inner error";
         case knowhere::Status::invalid_binary_set:
             return "invalid binary set";
+        case knowhere::Status::invalid_instruction_set:
+            return "current index not supported in this cpu";
```
Comment (alexanderguzhva) [inline include/knowhere/expected.h:85]: "the current CPU model lacks hardware instructions that are needed for the implementation of the chosen index type"

File: src/index/ivf/ivf_config.h [hunk 2]
```diff
@@ -75,6 +76,11 @@ class ScannConfig : public IvfFlatConfig {
 
     inline Status
     CheckAndAdjustForSearch(std::string* err_msg) override {
+        if (!faiss::support_pq_fast_scan) {
+            *err_msg = "scann index can not be used on current cpu, need at least avx2 instruction set";
```
Comment (alexanderguzhva) [inline src/index/ivf/ivf_config.h:80]: "the current index is not supported on the current CPU model. AVX2 support is needed for x86."

File: src/index/ivf/ivf_config.h [hunk 3]
```diff
@@ -86,6 +92,25 @@ class ScannConfig : public IvfFlatConfig {
 
         return Status::success;
     }
+
+    inline Status
+    CheckAndAdjustForRangeSearch(std::string* err_msg) override {
+        if (!faiss::support_pq_fast_scan) {
+            *err_msg = "scann index can not be used on current cpu, need at least avx2 instruction set";
```
Comment (alexanderguzhva) [inline src/index/ivf/ivf_config.h:99]: "the current index is not supported on the current CPU model. AVX2 support is needed for x86."

File: src/index/ivf/ivf_config.h [hunk 4]
```diff
@@ -86,6 +92,25 @@ class ScannConfig : public IvfFlatConfig {
...
+            LOG_KNOWHERE_ERROR_ << *err_msg;
+            return Status::invalid_instruction_set;
+        }
+        return Status::success;
+    }
+
+    inline Status
+    CheckAndAdjustForBuild() override {
+        if (!faiss::support_pq_fast_scan) {
+            LOG_KNOWHERE_ERROR_ << "scann index can not be used on current cpu, need at least avx2 instruction set";
```
Comment (alexanderguzhva) [inline src/index/ivf/ivf_config.h:109]: "the current index is not supported on the current CPU model. AVX2 support is needed for x86."

File: cmake/libs/libfaiss.cmake [hunk 5]
```diff
@@ -52,6 +57,16 @@ endif()
...
 if(__X86_64)
+  list(REMOVE_ITEM FAISS_SRCS ${FAISS_AVX2_SRCS})
+
+  add_library(faiss_avx2 OBJECT ${FAISS_AVX2_SRCS})
+  target_compile_options(
+    faiss_avx2
+    PRIVATE $<$<COMPILE_LANGUAGE:CXX>:
+            -msse4.2
+            -mavx2
+            -mfma
```
Comment (alexanderguzhva) [inline cmake/libs/libfaiss.cmake:68]: also `-mpopcnt`, please :)

File: src/index/ivf/ivf_config.h (hunk 2 above)
Comment (chasingegg) [reply src/index/ivf/ivf_config.h:80]: This check is for one specific index so we can directly point out the index name here, and `AVX2 support is needed for x86` is good indeed

**Comment by liliu-z** [comment]
> 1. It is not IndexScaNN being unsupported, it is PQ fast scan procedure being unsupported. Technically, there are multiple indices in Faiss that rely on PQ fast scan, so I would change `support_scann` variable name to `support_pq_fast_scan` to make it clear and acceptable for future possible indices without refactoring.
> 2. Intel(R) Xeon(R) CPU E5-2420 supports AVX, but not AVX2, which is written in the flags in [[Bug]: Milvus2.3.1 standalone exited with code 132 milvus-io/milvus#27552](https://github.com/milvus-io/milvus/issues/27552) and also mentioned on the [manufacturer web page](https://www.intel.com/content/www/us/en/products/sku/64617/intel-xeon-processor-e52420-15m-cache-1-90-ghz-7-20-gts-intel-qpi/specifications.html) . So, I would transition the change into `Make knowhere run on old cpus which only support AVX`, if possible

About the second points, Knowhere needs to be runnable on all CPUs not only the one with AVX and more

File: include/knowhere/config.h [hunk 6]
```diff
@@ -564,7 +564,7 @@ class BaseConfig : public Config {
     }
 
     virtual Status
-    CheckAndAdjustForRangeSearch() {
+    CheckAndAdjustForRangeSearch(std::string* err_msg) {
```
Comment (liliu-z) [inline include/knowhere/config.h:567]: I didn't see anywhere you use this msg

Comment (chasingegg) [reply include/knowhere/config.h:567]: In scann `CheckAndAdjustForRangeSearch` will return err_msg if cpu not supported, similar with `CheckAndAdjustForSearch`

**Pull request merged**